src/psbt.c

```diff
@@ -0,0 +1,2120 @@
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct wally_psbt);
+
+    if (inputs_allocation_len) {
+        new_inputs = wally_malloc(inputs_allocation_len * sizeof(struct wally_psbt_input));
+        wally_bzero(new_inputs, inputs_allocation_len * sizeof(struct wally_psbt_input));
+    }
+    if (outputs_allocation_len) {
+        new_outputs = wally_malloc(outputs_allocation_len * sizeof(struct wally_psbt_output));
+        wally_bzero(new_outputs, outputs_allocation_len * sizeof(struct wally_psbt_output));
+    }
+    unknowns_map_init_alloc(global_unknowns_allocation_len, &result->unknowns);
+    if ((inputs_allocation_len && !new_inputs) ||
+        (outputs_allocation_len && !new_outputs) ||
+        (global_unknowns_allocation_len && !result->unknowns)) {
+        wally_free(new_inputs);
+        wally_free(new_outputs);
+        wally_free(result->unknowns);
+        wally_free(result);
+        *output = NULL;
+        return WALLY_ENOMEM;
+    }
+
+    result->inputs = new_inputs;
+    result->num_inputs = 0;
+    result->inputs_allocation_len = inputs_allocation_len;
+    result->outputs = new_outputs;
+    result->num_outputs = 0;
+    result->outputs_allocation_len = outputs_allocation_len;
+    result->tx = NULL;
+    return WALLY_OK;
+}
+
+int wally_psbt_free(struct wally_psbt *psbt)
+{
+    size_t i;
+    if (psbt) {
+        wally_tx_free(psbt->tx);
+        for (i = 0; i < psbt->num_inputs; ++i) {
+            wally_psbt_input_free(&psbt->inputs[i]);
+        }
+        for (i = 0; i < psbt->num_outputs; ++i) {
+            wally_psbt_output_free(&psbt->outputs[i]);
+        }
+        unknowns_map_free(psbt->unknowns);
+    }
+    return WALLY_OK;
+}
+
+int wally_psbt_set_global_tx(
+    struct wally_tx *tx,
+    struct wally_psbt *psbt)
+{
+    int ret = WALLY_OK;
+    size_t i;
+
+    if ((ret = clone_tx(tx, &psbt->tx)) != WALLY_OK) {
+        return ret;
+    }
+
+    for (i = 0; i < psbt->num_inputs; ++i) {
+        wally_psbt_input_free(&psbt->inputs[i]);
+    }
+    if (psbt->inputs_allocation_len < tx->num_inputs) {
+        wally_free(psbt->inputs);
+        psbt->inputs_allocation_len = 0;
+        psbt->inputs = wally_malloc(tx->num_inputs * sizeof(struct wally_psbt_input));
+        if (!psbt->inputs) {
+            return WALLY_ENOMEM;
+        }
+        psbt->inputs_allocation_len = tx->num_inputs;
+    }
+    wally_bzero(psbt->inputs, psbt->inputs_allocation_len * sizeof(struct wally_psbt_input));
+    psbt->num_inputs = tx->num_inputs;
+
+    for (i = 0; i < psbt->num_outputs; ++i) {
+        wally_psbt_output_free(&psbt->outputs[i]);
+    }
+    if (psbt->outputs_allocation_len < tx->num_outputs) {
+        wally_free(psbt->outputs);
+        psbt->outputs_allocation_len = 0;
+        psbt->outputs = wally_malloc(tx->num_outputs * sizeof(struct wally_psbt_input));
+        psbt->outputs_allocation_len = tx->num_outputs;
+        if (!psbt->outputs) {
+            return WALLY_ENOMEM;
```

Comment:
How so?

Reply:
same as above or leaves them allocated as a side effect.

Reply:
I've changed this so that it only works on a newly allocated psbt. So there will be no inputs or outputs. If one of these allocations fails, the other one doesn't need to be undone as it is just an allocation of an empty array.